Feature: Add cors configuration

SportWizard opened the pull request:
This PR configures CORS and some added packages

Settings:
- origin: If the backend server is in "development" (using npm run dev), it only allows "http://localhost:4200" (the frontend server) to send requests to the backend server. Otherwise, it only allows "https://https://whcfc.ca" to send requests to the backend server
- allowedHeader: Only request headers with "Content-Type" are permitted (e.g. "application/json", "application/xml", "text/plain“)

Packages:
- nodemon: Automatically restarts a Node.js app when source files change so you don’t need to stop/start the server manually
- cross-env: Provide a consistent way to set env vars in package.json scripts (avoids Windows vs POSIX differences)

SophieShen897 commented:
Thank you for your contribution! This PR has taken an important step in the right direction, and the overall architectural thinking is excellent. However, a few key issues need to be fixed before merging.

1. HTTP Header Case Sensitivity Error
Location: WHCFC_Backend/index.js Line 6
Explanation: Although HTTP headers are case-insensitive, the standard format should be Content-Type (with a capital T). This could potentially cause issues with some strict clients.

2. Missing NODE_ENV Setting in Production Environment
Location: WHCFC_Backend/package.json Line 7

Issue: The current start script does not set NODE_ENV, which can lead to:
Environment checks in the CORS configuration not functioning correctly
The selection of environment variable files in index.js (Line 9) not working as expected. If the NODE_ENV is not specified, you won't know which configuration file it will load. production or development

3. Incomplete CORS Configuration
Location: WHCFC_Backend/index.js Lines 15-17

The current CORS configuration is too simplistic and should be enhanced.

3.1 The current configuration only allows:

"https://whcfc.ca"
However, users might access the site through various URLs, including:

"https://www.whcfc.ca" // with www
"https://whcfc.ca" // without www
3.2 There are no specified allowed HTTP methods.
The backend handles both GET and POST requests, and explicitly specifying these can prevent unintended methods (like PUT and DELETE) from being allowed.

3.3 Add maxAge for optimization.
When a browser first sends a cross-origin request, it may send a "preflight request" (OPTIONS). Setting maxAge tells the browser not to send preflight requests again for a specified duration (e.g., 24 hours), thus reducing unnecessary network requests.

Regarding the dependency issue of cross-env
It works fine in the development environment, but is it problematic in the production environment?

SportWizard commented:
Hi Sophie. I have done following fixes
- Convert "Content-type" to "Content-Type"
- Added multiple origins
- Allow only Get and POST requests
- Set maxAge to 3600s (1 hour)
- Removed cross-env because it didn't need for crosConfig and the loading of env file is already done in the package.json

Thank you for your suggestions.

SophieShen897 commented:
Thanks for the quick updates! However, I've discovered a new critical security issue that needs immediate attention.

We currently have a serious problem with allowing localhost access in the production environment. In the `WHCFC_Backend/index.js` file on line 16, the CORS configuration includes `http://localhost:4200`, which is extremely dangerous.

Here’s why: if an attacker runs a malicious program on a user’s computer, they could start a server on `localhost:4200`, trick the user into visiting a harmful webpage, and then directly call your production API. Since CORS allows requests from `localhost:4200`, this could succeed. 

Additionally, if an attacker gains control over someone’s development environment, they could access the production API from `localhost:4200` and bypass all the CORS protections. The root of the issue is that you commented out the environment checks, resulting in the same configuration being used across all environments. It would be better to use environment variables or validation functions to handle different settings appropriately.

Regarding the commented-out Dotenv code, I noticed that you have `import "dotenv/config"` at the start of the file, which automatically loads the `.env` file. Commenting this out may not cause issues, but it prevents automatic switching between `.env.dev` and `.env` for different environments. This means the development environment could unintentionally use production settings or vice versa.

To set up `NODE_ENV`, you can configure it like this for the development environment in `package.json`:
```json
{
  "scripts": {
    "dev": "set NODE_ENV=development && npx @dotenvx/dotenvx run -f .env.dev -- nodemon index.js"
  }
}
```
For production, make sure to set the environment variable on your deployment platform (like Heroku, AWS, Netlify, etc.):
```bash
NODE_ENV=production
```
Or you can specify it in the startup command:
```bash
NODE_ENV=production node index.js
```
This review was stringent because CORS configuration is crucial for API security. Thank you for the deduction.
